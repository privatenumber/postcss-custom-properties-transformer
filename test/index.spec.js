const {transform, md5} = require('./utils');
const fixtures = require('./fixtures');

describe('error handling', () => {
	test('no options', () => {
		expect(() => transform(fixtures)).toThrowError('[postcss-custom-properties-transformer] a transformer must be passed in');
	});

	test('no transformer', () => {
		expect(() => transform(fixtures, {})).toThrowError('[postcss-custom-properties-transformer] a transformer must be passed in');
	});

	test('invalid transformer', () => {
		expect(() => transform(fixtures, {
			transformer: 123,
		})).toThrowError('[postcss-custom-properties-transformer] transformer must be a function');
	});

	test('transformer collision warning', async () => {
		const spy = jest.spyOn(global.console, 'warn').mockImplementation();
		await transform(fixtures, {
			transformer() {
				return 'constant-value';
			},
		});
		expect(spy).toHaveBeenCalledWith('[postcss-custom-properties-transformer] Collision: property name "--constant-value" was generated from input "--font-size" but was already generated from input "--color"');
		spy.mockRestore();
	});
});

describe('transformer function', () => {
	test('namespace', async () => {
		const output = await transform(fixtures, {
			transformer({property}) {
				return `app-${property}`;
			},
		});

		expect(output).toMatchSnapshot();
	});

	test('emoji namespace', async () => {
		const output = await transform(fixtures, {
			transformer({property}) {
				return `ðŸ”¥-${property}`;
			},
		});

		expect(output).toMatchSnapshot();
	});

	test('custom hash', async () => {
		const output = await transform(fixtures, {
			transformer({property}) {
				return md5(property).slice(0, 6);
			},
		});

		expect(output).toMatchSnapshot();
	});

	test('custom hash - scoped to file', async () => {
		const output = await transform(fixtures, {
			transformer({property, filepath}) {
				return md5(filepath + property).slice(0, 6);
			},
		});

		expect(output).toMatchSnapshot();
	});
});

describe('transform options', () => {
	test('value declarations', async () => {
		const output = await transform(fixtures, {
			transformValueDeclarations: true,
			transformer({property}) {
				return `app-${property}`;
			},
		});

		expect(output).toMatchSnapshot();
	});
});
