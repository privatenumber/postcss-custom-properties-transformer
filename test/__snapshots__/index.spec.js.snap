// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer function custom hash - scoped to file 1`] = `
Array [
  "
	.button {
		--ae7621: red;
		--b682a4: 24px;
		background: var(--ae7621);
		border-color: var(--ae7621);
		font-size: var(--b682a4);
	}

	.input {
		--ae7621: red;
		--b682a4: 24px;
		background: var(--ae7621);
		border-color: var(--ae7621);
		font-size: var(--b682a4);
		border: 1px solid var(--ae7621);
	}
	",
  "
	:root {
		--199650: red;
		--29d9c5: 24px;
		--d85b32: calc(var(--29d9c5) * 2);
	}
	.section {
		background: var(--199650);
		border-color: var(--199650);
		font-size: var(--29d9c5);
	}

	.heading {
		background: var(--199650);
		border-color: var(--199650);
		font-size: var(--29d9c5);
		border: 1px solid var(--199650);
	}
	",
]
`;

exports[`transformer function custom hash 1`] = `
Array [
  "
	.button {
		--70dda5: red;
		--9b1d0a: 24px;
		background: var(--70dda5);
		border-color: var(--70dda5);
		font-size: var(--9b1d0a);
	}

	.input {
		--70dda5: red;
		--9b1d0a: 24px;
		background: var(--70dda5);
		border-color: var(--70dda5);
		font-size: var(--9b1d0a);
		border: 1px solid var(--70dda5);
	}
	",
  "
	:root {
		--70dda5: red;
		--9b1d0a: 24px;
		--defe4c: calc(var(--9b1d0a) * 2);
	}
	.section {
		background: var(--70dda5);
		border-color: var(--70dda5);
		font-size: var(--9b1d0a);
	}

	.heading {
		background: var(--70dda5);
		border-color: var(--70dda5);
		font-size: var(--9b1d0a);
		border: 1px solid var(--70dda5);
	}
	",
]
`;

exports[`transformer function emoji namespace 1`] = `
Array [
  "
	.button {
		--ðŸ”¥-color: red;
		--ðŸ”¥-font-size: 24px;
		background: var(--ðŸ”¥-color);
		border-color: var(--ðŸ”¥-color);
		font-size: var(--ðŸ”¥-font-size);
	}

	.input {
		--ðŸ”¥-color: red;
		--ðŸ”¥-font-size: 24px;
		background: var(--ðŸ”¥-color);
		border-color: var(--ðŸ”¥-color);
		font-size: var(--ðŸ”¥-font-size);
		border: 1px solid var(--ðŸ”¥-color);
	}
	",
  "
	:root {
		--ðŸ”¥-color: red;
		--ðŸ”¥-font-size: 24px;
		--ðŸ”¥-large-font-size: calc(var(--ðŸ”¥-font-size) * 2);
	}
	.section {
		background: var(--ðŸ”¥-color);
		border-color: var(--ðŸ”¥-color);
		font-size: var(--ðŸ”¥-font-size);
	}

	.heading {
		background: var(--ðŸ”¥-color);
		border-color: var(--ðŸ”¥-color);
		font-size: var(--ðŸ”¥-font-size);
		border: 1px solid var(--ðŸ”¥-color);
	}
	",
]
`;

exports[`transformer function namespace 1`] = `
Array [
  "
	.button {
		--app-color: red;
		--app-font-size: 24px;
		background: var(--app-color);
		border-color: var(--app-color);
		font-size: var(--app-font-size);
	}

	.input {
		--app-color: red;
		--app-font-size: 24px;
		background: var(--app-color);
		border-color: var(--app-color);
		font-size: var(--app-font-size);
		border: 1px solid var(--app-color);
	}
	",
  "
	:root {
		--app-color: red;
		--app-font-size: 24px;
		--app-large-font-size: calc(var(--app-font-size) * 2);
	}
	.section {
		background: var(--app-color);
		border-color: var(--app-color);
		font-size: var(--app-font-size);
	}

	.heading {
		background: var(--app-color);
		border-color: var(--app-color);
		font-size: var(--app-font-size);
		border: 1px solid var(--app-color);
	}
	",
]
`;
